src/index.js

```diff
@@ -1,169 +1,323 @@
+	 * @param {PopStateEvent} popstate
+	 */
+	onTransitionStart = (popstate) => {
+		if (this.options.doScrollingRightAway && !this.swup.scrollToElement) {
+			this.doScrollingBetweenPages(popstate);
+		}
+	};
+
+	/**
+	 * Handles `contentReplaced`
+	 * @param {PopStateEvent} popstate
+	 */
+	onContentReplaced = (popstate) => {
+		if (!this.options.doScrollingRightAway || this.swup.scrollToElement) {
+			this.doScrollingBetweenPages(popstate);
+		}
+
+		this.restoreScrollContainers(popstate);
+	};
+
+	/**
+	 * Scrolls the window, based on context
+	 *
+	 * @param {PopStateEvent|boolean} popstate
+	 * @returns void
+	 */
+	doScrollingBetweenPages = (popstate) => {
+		const swup = this.swup;
+
+		// Bail early on popstate and inactive `animateHistoryBrowsing`
+		if (popstate && !swup.options.animateHistoryBrowsing) {
+			return;
+		}
+
+		// Try scrolling to a given anchor
+		if (this.maybeScrollToAnchor(swup.scrollToElement, 'betweenPages')) {
+			swup.scrollToElement = null;
+			return;
+		}
+
+		// Finally, scroll to either the stored scroll position or to the very top of the page
+		const scrollPositions = this.getStoredScrollPositions(getCurrentUrl());
+		const top = (scrollPositions.window && scrollPositions.window.top) || 0;
+		swup.scrollTo(top, this.shouldAnimate('betweenPages'));
+	};
+
+	/**
+	 * Stores the current scroll positions for the URL we just came from
+	 */
+	onWillReplaceContent = () => {
+		this.storeScrollPositions(this.previousUrl);
+		this.previousUrl = getCurrentUrl();
+	};
+
+	/**
+	 * Deletes the scroll positions for the URL a link is pointing to,
+	 * if shouldRestoreScrollPosition doesn't evaluate to true
+	 *
+	 * @param {event} e
+	 * @returns void
+	 */
+	onClickLink = (e) => {
+		if (!this.options.shouldRestoreScrollPosition(e.delegateTarget)) {
+			return;
+		}
+		const url = new Link(e.delegateTarget).getAddress();
+		this.deleteStoredScrollPositions(url);
+	};
+
+	/**
+	 * Stores the scroll positions for the current URL
+	 * @param {string} url
+	 */
+	storeScrollPositions(url) {
+		// create an object to store the scroll positions
+		const storeEntry = {
+			window: { top: window.scrollY, left: window.scrollX },
+			containers: []
+		};
+		// fill up the object with scroll positions for each matching element
+		const $containers = document.querySelectorAll(this.options.scrollContainers);
+		$containers.forEach((el) =>
+			storeEntry.containers.push({
+				top: el.scrollTop,
+				left: el.scrollLeft
+			})
+		);
+		// put the object into the store
+		this.scrollPositionsStore[url] = storeEntry;
+	}
+
+	/**
+	 * Deletes stored scroll positions for a given URL
+	 * @param {string} url
+	 */
+	deleteStoredScrollPositions(url) {
+		delete this.scrollPositionsStore[url];
+		this.scrollPositionsStore[url] = null;
+	}
+
+	/**
+	 * Get the stored scroll positions for a given URL from the cache
+	 * @returns {object}
+	 */
+	getStoredScrollPositions(url) {
+		return this.scrollPositionsStore[url] || {};
```

Comment:
I'd suggest not returning the fallback object `{}` here. It makes it easier to check for a missing entry elsewhere and it's more obvious what's happening. The check on line 309/310 (`scrollPositions.containers == null`) would need to change then to check for null instead of assuming an empty object.  But as a new contributor it'd be easier to grok I think.

Reply:
If we did this, we would have to do `scrollPositions && scrollPositions.containers`, also the same for `scrollPositions.window` in on [line 237](https://github.com/hirasso/scroll-plugin/blob/feature/restore-scroll/src/index.js#L237). That's why I would prefer the function to always return an object.

Reply:
You could move the `|| {}` fallback to after the method call. That'd make it just as simple. Returning an empty object is a bit of a code smell as you need to reach into the empty object to even know if you got a return.